stlab/future.hpp

```diff
@@ -814,23 +814,24 @@ struct when_all_shared {
     std::tuple<boost::optional<Ts>...>  _args;
     future<void>                        _holds[sizeof...(Ts)] {};
     std::atomic_size_t                  _remaining {sizeof...(Ts)};
-    std::atomic_bool                    _error_happend{ false };
-    std::mutex                          _errormutex;
+    std::atomic_bool                    _error_happened{ false };
     boost::optional<std::exception_ptr> _error;
     packaged_task<>                     _f;
 
-    void done() { if (--_remaining == 0 && !_error_happend) _f(); }
+    void done() { if (!_error_happened && --_remaining == 0) _f(); }
 
     void failure(std::exception_ptr error) {
-        {
-            std::unique_lock<std::mutex> lock(_errormutex);
+        bool current_error_happened = _error_happened.load();
```

Comment:
This is an unnecessary check.

if (_error_happened.compare_exchange_strong(false, true)) {
  _error = std::move(error);
  _f();
}